src/meta/meta_ckpt.c

```diff
@@ -1262,6 +1262,195 @@ __wt_meta_ckptlist_update_config(
+{
+    size_t index;
+    uint8_t *new_ptr, *original_ptr;
+    uint8_t partial_result;
+
+    WT_NOT_READ(index, 0);
+    original_ptr = NULL;
+    new_ptr = NULL;
+    WT_NOT_READ(partial_result, 0);
+
+    *ok = false;
+
+    if ((original_bitmap == NULL) || (new_bitmap == NULL))
+        return (EINVAL);
+
+    /* If the new bitmap is smaller, then some bits have been lost which is a problem. */
+    if (original_bitmap->size > new_bitmap->size)
+        return (EINVAL);
+
+    original_ptr = (uint8_t *)original_bitmap->data;
+    new_ptr = (uint8_t *)new_bitmap->data;
+
+    *ok = true;
+
+    for (index = 0; index < original_bitmap->size; index++) {
+        /*
+         * Detect bits that were (incorrectly) changed from 1 in the original to 0 in the new.
+         *
+         * This check is performed by bitwise inverting (using bitwise NOT) the new value
+         * so that any bits that are, or have become 0, are turned into a 1. This intermediate
+         * value is then bitwise ANDed with the original value, so that any bit values that are 1 in
+         * the original result and 0 in the new value (and therefore 1 in the intermediate value)
+         * results in a bit that is set in the partial result. Any 0s in the original value result
+         * in 0 in the partial result.
+         *
+         * Thus, any non-zero partial results indicate a bit that has incorrectly changed from
+         * 1 in the original to 0 in the new.
+         */
+        partial_result = *original_ptr & ~(*new_ptr);
+
+        if (partial_result != 0) {
+            *ok = false;
+            break;
+        }
+
+        original_ptr++;
+        new_ptr++;
+    }
+
+    return (0);
+}
+
+/*
+ * __check_backup_blocks --
+ *     This function checks that the backup blocks for a checkpoint are correct, and that there are
+ *     no errors in the incremental backup blkmod information.
+ */
+static WT_INLINE int
+__check_backup_blocks(
+  WT_SESSION_IMPL *session, WT_CKPT *ckpt, WT_ITEM *checkpoint_blkmods_buffer, const char *filename)
+{
+    WT_BLOCK_MODS *blk;
+    WT_DECL_RET;
+    WT_ITEM file_blkmods_buffer;
+    u_int i;
+    bool blkmods_are_ok;
+
+    blkmods_are_ok = true;
+    WT_CLEAR(file_blkmods_buffer);
+
+    for (i = 0, blk = &ckpt->backup_blocks[0]; i < WT_BLKINCR_MAX; ++i, ++blk) {
+        if (!F_ISSET(blk, WT_BLOCK_MODS_VALID))
+            continue;
+
+        WT_ERR(__ckpt_extract_blkmod_bitmap(session, ckpt, blk->id_str, checkpoint_blkmods_buffer));
+
+        WT_ERR(__get_blkmods(session, filename, blk->id_str, &file_blkmods_buffer));
+        if (checkpoint_blkmods_buffer->size > 0) {
+            if (file_blkmods_buffer.size > 0) {
+                blkmods_are_ok = false;
+                ret = __check_incorrect_modified_bits(
+                  &file_blkmods_buffer, checkpoint_blkmods_buffer, &blkmods_are_ok);
+
+                if ((ret != 0) || !blkmods_are_ok) {
+                    WT_ASSERT(session, false);
```

Comment:
Why not a `WT_ASSERT_ALWAYS` instead?

Reply:
We need it because otherwise we get the following errors from the clang analyser:

```
[2024/03/17 19:54:29.018] /data/mci/8e9fbae01b34d2314f0ee58c6aa5e59e/wiredtiger/src/meta/meta_ckpt.c:1362:5: warning: Value stored to 'index' is never read [deadcode.DeadStores]
[2024/03/17 19:54:29.018]     index = 0;
[2024/03/17 19:54:29.018]     ^       ~
[2024/03/17 19:54:29.018] /data/mci/8e9fbae01b34d2314f0ee58c6aa5e59e/wiredtiger/src/meta/meta_ckpt.c:1365:5: warning: Value stored to 'partial_result' is never read [deadcode.DeadStores]
[2024/03/17 19:54:29.018]     partial_result = 0;
```

Reply:
I'm happy change it. Because of the subsequent `WT_ERR_PANIC()`, I think either `WT_ASSERT` or `WT_ASSERT_ALWAYS` is ok here.

Reply:
Should we keep both though?

Reply:
I think the panic is the right option and I'm ok with having either the WT_ASSERT or the WT_ASSERT_ALWAYS with it. I think there was a comment suggesting the WT_ASSERT be added much earlier in the discussion, but I don't seem to be able to find it now.